packages/core/src/server.jsx

```diff
@@ -147,6 +147,29 @@ export default function(defaultStore = appInitialState, headerConfig, reduxMiddl
             preRenderMiddleware(store, newProps)
               .then(() => {
 
+                // Needs may return a special canonRedirect key. If they do so, process a redirect, using the variables provided
+                // in those objects as variables to substitute in the routes.
+                const redirects = Object.values(store.getState().data).filter(d => d.canonRedirect);
+                // If the query contains ?redirect=true, a redirect has already occurred. To avoid redirect loops, ensure this value is unset
+                if (!req.query.redirect && redirects.length > 0) {
+                  // If any needs provided redirect keys, combine them into one object.
+                  const variables = redirects.reduce((acc, d) => ({...acc, ...d.canonRedirect}), {});
+                  // Use variables given by the canonRedirect key, but fall back on given params (to cover for unprovided keys, like :lang)
+                  const params = {...props.params, ...variables};
+                  // Not sure if this is a reliable way to get which route this is.
+                  let route = props.routes[1].path;
```

Comment:
I believe this is correct, because `props` is coming from the react-router `match` function.